### PR TITLE
Fix PoseidonConstants `serde` Implementation

### PR DESCRIPTION
**Pull request by winston-h-zhang**
Addresses https://github.com/lurk-lab/lurk-rs/issues/401.

From @huitseeker's insight, the manual implementation had some alignment issues. The solution was to directly `#[derive(Serialize, Deserialize)]` for `PoseidonConstants`, as well as add the `#[serde(bound)]` to generate a correct implementation.

### CODE REVIEW

**Review by porcuquine — APPROVED** [review]

File: Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ log = "0.4.17"
 pasta_curves = { version = "0.5", features = ["serde"] }
 trait-set = "0.3.0"
 serde = { version = "1.0", features = ["derive"] }
+bincode = "1.3.3"
```
Comment (huitseeker) [inline Cargo.toml:25]: It seems you should be able to add bincode only to dev-dependencies

**Review by huitseeker — COMMENTED** [review]
Re:CI, see https://github.com/lurk-lab/lurk-rs/pull/356

**Review by huitseeker — APPROVED** [review]

**Review by huitseeker — CHANGES_REQUESTED** [review]

File: src/serde_impl.rs [hunk 2]
```diff
@@ -1,287 +0,0 @@
...
-        const FIELDS: &[&str] = &["
-	  mds_matrices,
-	  compressed_round_constants,
-	  pre_sparse_matrix,
-	  sparse_matrixes,
-	  strength,
-	  full_rounds,
-	  partial_rounds,
-	  hash_type,
-"];
```
Comment (huitseeker) [inline src/serde_impl.rs:232]: This was the original issue with the fields in this custom impl, BTW. Note the double quotes, the slice of length 1, and how this is declaring a single-field struct.

File: src/poseidon.rs [hunk 3]
```diff
@@ -1309,7 +1309,8 @@ mod tests {
         let mut h3 = Poseidon::<Fr, U2>::new_with_preimage(&preimage, &constants3);
 
         assert_eq!(h1.hash(), h2.hash());
-        assert_eq!(h1.hash(), h3.hash())
+        h1.set_preimage(&preimage); // reset
```
Comment (huitseeker) [inline src/poseidon.rs:1312]: 🤦 hahaha, got it, thanks for the fix

**Review by huitseeker — APPROVED** [review]
Thanks a lot! I've checked the generated code is the same as the original, except for correcting the originally-mangled FIELDS constant.

**Comment by winston-h-zhang** [comment]
On `Nova`, I'm running into a very strange error:
```rust
running 1 test
thread 'provider::poseidon::tests::test_serde_roundtrip_what' panicked at 'called `Result::unwrap()` on an `Err` value: 
Custom("invalid value: integer `6`, expected variant index 0 <= i < 6")', src/provider/poseidon.rs:291:84
```

Looking around, I found https://github.com/bincode-org/bincode/issues/424, so it seems like using `serde(skip)` is unstable. I'm not sure why, but the roundtrip tests pass on Neptune and fail on Nova... I've been banging my head on this for the past hour to no avail. In any case, we may have to revert back to Sam's custom implementation.

**Comment by winston-h-zhang** [comment]
Ok, with this, we should finally be out of the thorns @huitseeker @samuelburnham!

File: src/hash_type.rs [hunk 4]
```diff
@@ -12,20 +12,25 @@
...
     VariableLength,
     ConstantLength(usize),
     Encryption,
-    #[serde(skip)]
+    // #[serde(skip)] // See: https://github.com/bincode-org/bincode/issues/424
     Custom(CType<F, A>),
     Sponge,
 }
 
+#[doc(hidden)]
```
Comment (huitseeker) [inline src/hash_type.rs:33]: Why?

File: src/serde_impl.rs [hunk 5]
```diff
@@ -1,287 +0,0 @@
...
-                Ok(PoseidonConstants {
-                    mds_matrices,
-                    round_constants: None,
-                    compressed_round_constants,
-                    pre_sparse_matrix,
-                    sparse_matrixes,
-                    strength,
-                    domain_tag: hash_type.domain_tag(),
-                    full_rounds,
-                    half_full_rounds: full_rounds / 2,
```
Comment (huitseeker) [inline src/serde_impl.rs:116]: This line is valuable

File: src/serde_impl.rs [hunk 6]
```diff
@@ -1,287 +0,0 @@
...
-                let hash_type: HashType<F, A> =
-                    hash_type.ok_or_else(|| de::Error::missing_field("hash_type"))?;
-                Ok(PoseidonConstants {
-                    mds_matrices,
-                    round_constants: None,
-                    compressed_round_constants,
-                    pre_sparse_matrix,
-                    sparse_matrixes,
-                    strength,
-                    domain_tag: hash_type.domain_tag(),
```
Comment (huitseeker) [inline src/serde_impl.rs:213]: This is valuable

File: src/serde_impl.rs [hunk 7]
```diff
@@ -1,287 +0,0 @@
...
-                    full_rounds,
-                    half_full_rounds: full_rounds / 2,
```
Comment (huitseeker) [inline src/serde_impl.rs:215]: This is valuable

File: src/serde_impl.rs [hunk 8]
```diff
@@ -1,287 +0,0 @@
...
-                let partial_rounds = seq
-                    .next_element()?
-                    .ok_or_else(|| de::Error::invalid_length(6, &self))?;
-                let hash_type: HashType<F, A> = seq
-                    .next_element()?
-                    .ok_or_else(|| de::Error::invalid_length(7, &self))?;
-
-                Ok(PoseidonConstants {
-                    mds_matrices,
-                    round_constants: None,
```
Comment (huitseeker) [inline src/serde_impl.rs:109]: This is valuable

File: src/serde_impl.rs [hunk 9]
```diff
@@ -1,287 +0,0 @@
...
-                let strength = strength.ok_or_else(|| de::Error::missing_field("strength"))?;
-                let full_rounds =
-                    full_rounds.ok_or_else(|| de::Error::missing_field("full_rounds"))?;
-                let partial_rounds =
-                    partial_rounds.ok_or_else(|| de::Error::missing_field("partial_rounds"))?;
-                let hash_type: HashType<F, A> =
-                    hash_type.ok_or_else(|| de::Error::missing_field("hash_type"))?;
-                Ok(PoseidonConstants {
-                    mds_matrices,
-                    round_constants: None,
```
Comment (huitseeker) [inline src/serde_impl.rs:208]: This is valuable

**Review by huitseeker — CHANGES_REQUESTED** [review]
While I accepted a previous version, I think this version misses a few of the forced field implementations of the original (Deserialize for Poseidon constants). I've tried to comment online in the deleted impl where we lose behavior wrt the current. 
I would suggest:
- for round_constants, instead of serde_skip, we use [default](https://serde.rs/field-attrs.html#default)
- for the other fields, we can probably define them with [serde_with](https://serde.rs/field-attrs.html#with), though in aggregate, it may be easier to reuse the existing custom impl, and correct it's FIELDS constant?

**Comment by winston-h-zhang** [comment]
Ok, maybe it's best to just restore @samuelburnham's original impl, then

**Review by huitseeker — APPROVED** [review]
Thanks a lot @winston-h-zhang !

**Review by samuelburnham — APPROVED** [review]

**Pull request merged**